src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java

```diff
@@ -194,196 +180,483 @@ public StatementRestrictions addExternalRestrictions(Iterable<CustomIndexExpress
-         * WHERE clause. For a given entity, rules are:
-         *   - EQ relation conflicts with anything else (including a 2nd EQ)
-         *   - Can't have more than one LT(E) relation (resp. GT(E) relation)
-         *   - IN relation are restricted to row keys (for now) and conflicts with anything else (we could
-         *     allow two IN for the same entity but that doesn't seem very useful)
-         *   - The value_alias cannot be restricted in any way (we don't support wide rows with indexed value
-         *     in CQL so far)
-         */
-        for (Relation relation : whereClause.relations)
+            return doBuild(whereClause.root(), indexRegistry);
+        }
+
+        StatementRestrictions doBuild(WhereClause.ExpressionElement element, IndexRegistry indexRegistry)
         {
-            if (relation.operator() == Operator.IS_NOT)
-            {
-                if (!forView)
-                    throw invalidRequest("Unsupported restriction: %s", relation);
+            PartitionKeySingleRestrictionSet.Builder partitionKeyRestrictionSet = PartitionKeySingleRestrictionSet.builder(table.partitionKeyAsClusteringComparator());
+            ClusteringColumnRestrictions.Builder clusteringColumnsRestrictionSet = ClusteringColumnRestrictions.builder(table, allowFiltering, indexRegistry);
+            RestrictionSet.Builder nonPrimaryKeyRestrictionSet = RestrictionSet.builder();
+            ImmutableSet.Builder<ColumnMetadata> notNullColumnsBuilder = ImmutableSet.builder();
 
-                notNullColumnsBuilder.addAll(relation.toRestriction(table, boundNames).getColumnDefs());
-            }
-            else
+            for (Relation relation : element.relations())
             {
-                Restriction restriction = relation.toRestriction(table, boundNames);
+                if (relation.operator() == Operator.IS_NOT)
+                {
+                    if (!forView)
+                        throw invalidRequest("Unsupported restriction: %s", relation);
 
-                if (relation.isLIKE() && (!type.allowUseOfSecondaryIndices() || !restriction.hasSupportingIndex(indexRegistry)))
+                    notNullColumnsBuilder.addAll(relation.toRestriction(table, boundNames).getColumnDefs());
+                }
+                else
                 {
-                    if (getColumnsWithUnsupportedIndexRestrictions(table, ImmutableList.of(restriction)).isEmpty())
+                    Restriction restriction = relation.toRestriction(table, boundNames);
+
+                    if (relation.isLIKE() && (!type.allowUseOfSecondaryIndices() || !restriction.hasSupportingIndex(indexRegistry)))
+                    {
+                        if (getColumnsWithUnsupportedIndexRestrictions(table, ImmutableList.of(restriction)).isEmpty())
+                        {
+                            throw invalidRequest("LIKE restriction is only supported on properly indexed columns. %s is not valid.", relation.toString());
```

Comment:
What is the difference between `LIKE restriction is only supported on properly indexed columns.` and the next line `INDEX_DOES_NOT_SUPPORT_LIKE_MESSAGE`?

Reply:
This message is saying that the relation is against an unindexed column whereas the second message is saying that the column is indexed but the index doesn't support like.

Neither of these are particularly clear but they are existing messages so it's difficult to change them.